core/sourcehook/sourcehook.h

```diff
@@ -241,29 +273,242 @@ namespace SourceHook
+			memberFunc_ = other.memberFunc_;
+			other.staticFunc_ = nullptr;
+			other.memberFunc_ = nullptr;
+			return *this;
+		}
+
+		HookManagerPubFuncHandler& operator=(HookManagerPubFunc func)
+		{
+			staticFunc_ = func;
+			memberFunc_ = nullptr;
+			return *this;
+		}
+
+		HookManagerPubFuncHandler& operator=(IHookManagerMemberFunc* funcHandler)
+		{
+			staticFunc_ = nullptr;
+			memberFunc_ = funcHandler;
+			return *this;
+		}
+
+		operator bool() const
+		{
+			return (staticFunc_ != nullptr ||
+				memberFunc_ != nullptr);
+		}
+
+		operator HookManagerPubFunc() const
+		{
+			return staticFunc_;
+		}
+
+		operator IHookManagerMemberFunc*() const
+		{
+			return memberFunc_;
+		}
+
+		int operator()(bool store, IHookManagerInfo *hi) const
+		{
+			if(staticFunc_ != nullptr)
+				return staticFunc_(store, hi);
+
+			if(memberFunc_ != nullptr)
+				return memberFunc_->Call(store, hi);
+
+			return 0;
+		}
+
+		bool operator==(const HookManagerPubFuncHandler &rhs) const
+		{
+			return staticFunc_ == rhs.staticFunc_ &&
+			       memberFunc_ == rhs.memberFunc_;
+		}
+
+		bool operator==(HookManagerPubFunc staticFunc) const
+		{
+			return staticFunc_ == staticFunc &&
+			       memberFunc_ == nullptr;
+		}
+
+		bool operator==(const IHookManagerMemberFunc *memberFunc) const
+		{
+			return staticFunc_ == nullptr &&
+			       memberFunc_ == memberFunc;
+		}
+
+	private:
+		HookManagerPubFunc staticFunc_;
+		IHookManagerMemberFunc* memberFunc_;
+	};
+
+	class SHDelegateHandler;
+
 	class ISHDelegate
 	{
 	public:
 		virtual bool IsEqual(ISHDelegate *pOtherDeleg) = 0;		// pOtherDeleg is from the same plugin and hookman
+	private:
+		friend class SHDelegateHandler;
 		virtual void DeleteThis() = 0;
 	};
 
+	// ISHDelegate doesn't provide a virtual destructor, use this wrapper to handle it
+	class SHDelegateHandler final
+	{
+	public:
+		template<typename U, class ... Params>
+		static SHDelegateHandler Make(Params&&... params)
+		{
+			ISHDelegate* delegate = new U(std::forward<Params>(params)...);
+			return SHDelegateHandler(delegate);
+		}
+
+		SHDelegateHandler()
+		 : ptr_(nullptr, &SHDelegateHandler::DeleteDelegate)
+		{
+		}
+
+		explicit SHDelegateHandler(ISHDelegate* delegate)
+		 : ptr_(delegate, &SHDelegateHandler::DeleteDelegate)
+		{
+		}
+
+		void Reset()
+		{
+			ptr_.reset();
+		}
+
+		bool operator==(const SHDelegateHandler &other) const
+		{
+			return ptr_.get() == other.Get();
+		}
+
+		ISHDelegate* operator->() const
+		{
+			return ptr_.get();
+		}
+
+		ISHDelegate* Get() const
+		{
+			return ptr_.get();
+		}
+
+	private:
+		static void DeleteDelegate(ISHDelegate* delegate)
+		{
+			if(delegate)
+				delegate->DeleteThis();
+		}
+
+	private:
+		std::shared_ptr<ISHDelegate> ptr_;
+	};
+
 	struct IHookManagerInfo
 	{
 		virtual void SetInfo(int hookman_version, int vtbloffs, int vtblidx,
 			ProtoInfo *proto, void *hookfunc_vfnptr) = 0;
+
+		virtual void SetInfo(int hookman_version, int vtbloffs, int vtblidx,
```

Comment:
Overloads of virtual methods need to have a new name to avoid breaking compatibility on MSVC (it does a weird thing where it inverts the order). SetInfo2 is fine if you're feeling uncreative.